grails-app/services/org/pih/warehouse/core/DocumentTemplateService.groovy

```diff
@@ -0,0 +1,221 @@
+/**
+* Copyright (c) 2012 Partners In Health.  All rights reserved.
+* The use and distribution terms for this software are covered by the
+* Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
+* which can be found in the file epl-v10.html at the root of this distribution.
+* By using this software in any fashion, you are agreeing to be bound by
+* the terms of this license.
+* You must not remove this notice, or any other, from this software.
+**/
+package org.pih.warehouse.core
+
+import fr.opensagres.xdocreport.converter.ConverterTypeTo
+import fr.opensagres.xdocreport.converter.ConverterTypeVia
+import fr.opensagres.xdocreport.converter.Options
+import fr.opensagres.xdocreport.document.IXDocReport
+import fr.opensagres.xdocreport.document.registry.XDocReportRegistry
+import fr.opensagres.xdocreport.template.IContext
+import fr.opensagres.xdocreport.template.TemplateEngineKind
+import fr.opensagres.xdocreport.template.formatter.FieldsMetadata
+import groovy.text.Template
+import org.codehaus.groovy.grails.web.pages.GroovyPagesTemplateEngine
+import org.pih.warehouse.order.Order
+import org.pih.warehouse.order.OrderAdjustment
+import org.pih.warehouse.order.OrderItem
+import org.pih.warehouse.shipping.Shipment
+import org.pih.warehouse.shipping.ShipmentItem
+
+class DocumentTemplateService {
+
+    boolean transactional = true
+    GroovyPagesTemplateEngine groovyPagesTemplateEngine
+
+    def renderGroovyServerPageDocumentTemplate(Document documentTemplate, Map model) {
+        StringWriter output = new StringWriter()
+        String templateContents = new String(documentTemplate.fileContents)
+        Template template = groovyPagesTemplateEngine.createTemplate(templateContents, documentTemplate.name)
+        template.make(model).writeTo(output)
+        return output.toString()
+    }
+
+    def renderOrderDocumentTemplate(Document documentTemplate, Order orderInstance, ConverterTypeTo targetDocumentType, OutputStream outputStream) {
+
+        Boolean isVelocityTemplate = documentTemplate.filename.contains(".vm") || documentTemplate.filename.contains(".vtl")
+
+        TemplateEngineKind templateEngineKind = isVelocityTemplate ?
+                TemplateEngineKind.Velocity : TemplateEngineKind.Freemarker
+
+        InputStream inputStream = new ByteArrayInputStream(documentTemplate.fileContents)
+        IXDocReport report = XDocReportRegistry.getRegistry().loadReport(inputStream, templateEngineKind);
+
+        // FIXME Need a better way to handle this generically (consider using config + dataService)
+        IContext context = orderInstance ? createOrderContext(report, orderInstance) : report.createContext();
+
+        ConverterTypeVia sourceDocumentType = report.kind == "DOCX" ? ConverterTypeVia.DOCX4J :
+                report.getKind() == "ODT" ? ConverterTypeVia.ODFDOM : ConverterTypeVia.XWPF
```

Comment:
Maybe add brackets in here? I sometimes encounter issues because of it.

Reply:
@DominikaHatala Could you give me an example of what you mean?

Reply:
ConverterTypeVia sourceDocumentType = report.kind == "DOCX" ? ConverterTypeVia.DOCX4J :
                (report.getKind() == "ODT" ? ConverterTypeVia.ODFDOM : ConverterTypeVia.XWPF)

Reply:
I'm not sure if that is necessary in this case but I'm pretty sure that in the past I had wrong data when forgetting about brackets in nested ternary operator.

Reply:
Oh you meant parentheses. I could see something like this making it more readable. But I don't like the readability of the parantheses around the outer else.
```
ConverterTypeVia sourceDocumentType = (report.kind == "DOCX") ? ConverterTypeVia.DOCX4J :
(report.getKind() == "ODT") ? ConverterTypeVia.ODFDOM : ConverterTypeVia.XWPF
```